### ISSUE TITLE
Connection issues never seem to go away(how to deploy prisma live reliably) [MySQL, TypeGraphQL, EBS]

### ISSUE THREAD

**Issue by omar-dulaimi** [prisma/prisma#7128]
### Bug description

Hello,

Ever since our migration from Prisma 1 to Prisma 2, and connection issues never got solved for us. We wanted to deploy our same Nextjs website(after query changes) using the new Prisma 2 backend hosted on lambda using serverless framework and TypeGraphQL. 

We tried RDS proxy when it was first suggested by Prisma months ago, but then were surprised in the updated docs that it would be useless when used with Prisma.

That left us stuck, not knowing how to proceed from that. So we decided to step away from lambda and head to long running servers with elastic beanstalk. While preparing for that with the docker image, running the container locally using apollo server(not apollo server lambda), we still faced the same connection issues.

We referred back to the docs for the suggestions regarding deploying using the long running server model, we tried increasing the connection_limit from the default(9) to big numbers like 200 and 300. Connection issues were occurring just the same. We noticed that RDS connections never really go up to the number we specified while Prisma reports that it's actually using the number, while same issues were happening.

Since we're going the long-running server model, we want to guarantee that we won't face these issues once we go live. 

### How to reproduce

What we were testing the most was the homepage. In getServerSideProps(Nextjs), we prepare 5 graphql requests to get page content, and send them concurrently using Promise.all to the Prisma 2 backend.

### Expected behavior

The whole migration we did from Prisma 1 was in hopes of better database schema(Prisma 1 schema in the db level was a mess, many unnecessary relations, etc), better db access, reduced costs(no need for a separate server). But so far our experience has been horrible(deployment and testing stage), we kind of regretted doing the migration. We were better off with Prisma 1(despite Prisma 2 benefits).

### Prisma information

```js
const 1stPortion = services.getItems(
    user?.id,
    {
        ignoredBy: {
            none: {
                id: {
                    equals: user?.id ?? '',
                },
            },
        },
        status: {
            in: ['ACTIVE', 'ACTIVEUPDATED'],
        },
    },
    { score: 'desc' },
    4
);
const 2ndPortion = services.getItems(
    user?.id,
    {
        category: {
            equals: 'FARM',
        },
        type: {
            equals: 'RENT',
        },
        ignoredBy: {
            none: {
                id: {
                    equals: user?.id,
                },
            },
        },
        status: {
            in: ['ACTIVE', 'ACTIVEUPDATED'],
        },
    },
    { createdAt: 'desc' },
    4
);
const 3rdPortion = services.getItems(
    user?.id,
    {
        category: {
            equals: 'APARTMENT',
        },
        type: {
            equals: 'SALE',
        },
        ignoredBy: {
            none: {
                id: {
                    equals: user?.id,
                },
            },
        },
        status: {
            in: ['ACTIVE', 'ACTIVEUPDATED'],
        },
    },
    { createdAt: 'desc' },
    4
);
const 4thPortion = services.getItems(
    user?.id,
    {
        category: {
            equals: 'APARTMENT',
        },
        type: {
            equals: 'RENT',
        },
        ignoredBy: {
            none: {
                id: {
                    equals: user?.id,
                },
            },
        },
        status: {
            in: ['ACTIVE', 'ACTIVEUPDATED'],
        },
    },
    { createdAt: 'desc' },
    4
);
const 5thPortion = services.getItems(
    user?.id,
    {
        category: {
            equals: 'LAND',
        },
        type: {
            equals: 'SALE',
        },
        ignoredBy: {
            none: {
                id: {
                    equals: user?.id,
                },
            },
        },
        status: {
            in: ['ACTIVE', 'ACTIVEUPDATED'],
        },
    },
    { createdAt: 'desc' },
    4
);

const pageData = await Promise.all([
    1stPortion,
    2ndPortion,
    3rdPortion,
    4thPortion,
    5thPortion,
]); 
```

### Environment & setup

- OS: Ubuntu 18.04
- Database: MySQL 5.7
- Node.js version: v14.15.1
- Prisma version: 2.19

### Prisma Version

```
prisma               : 2.19.0
@prisma/client       : 2.18.0
Current platform     : debian-openssl-1.1.x
Query Engine         : query-engine c1455d0b443d66b0d9db9bcb1bb9ee0d5bbc511d (at ../../../../usr/local/lib/node_modules/prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine     : migration-engine-cli c1455d0b443d66b0d9db9bcb1bb9ee0d5bbc511d (at ../../../../usr/local/lib/node_modules/prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine : introspection-core c1455d0b443d66b0d9db9bcb1bb9ee0d5bbc511d (at ../../../../usr/local/lib/node_modules/prisma/node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary        : prisma-fmt c1455d0b443d66b0d9db9bcb1bb9ee0d5bbc511d (at ../../../../usr/local/lib/node_modules/prisma/node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Studio               : 0.358.0
```

**Comment by P4sca1** [comment]
What kind of connection issues are you referring to? Maybe we have the same issue.
I get the following errors after running my application for around a day: 
```
Error: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 10.
```
The issue does not appear during high load and prisma never recovers from this error. It just keeps spamming and an application restart is needed.
I raised the connection limit from 2 to 10 without success.

I checked database connections (MySQL 8.0) when the error occurs using `SHOW PROCESSLIST` and noticed that not a single connection is actually used.

**Comment by omar-dulaimi** [comment]
@P4sca1 Yeah, same issue. Changing connection limit doesn't really help.

**Comment by janpio** [comment]
@omar-dulaimi Thanks for the writeup here.

What is missing is the information how you get from the codesnippet you posted to an error message. With the information I have, I would assume sending a request that executes that code runs 5 queries in parallel, so will use 5 connections in parallel, and then returns that data to the request and user - easy.

But that might be a wrong assumption if `services.getItems` does more than to simply send a simple query. Can you share more here?

Also, when are you getting these connection errors - just by clicking around your app manually by hand, 1 page at a time, or while load testing with hundreds or thousands of virtual users? How is your database configured?

Besides these open questions:

> prisma               : 2.19.0
> @prisma/client       : 2.18.0

You should always use identical versions for CLI and Client. Please also try to upgrade to a more recent verison of both, 2.23.0 was released last week.

I am happy to debug this together with you. I will just need some more information from you.

**Comment by janpio** [comment]
@P4sca1 I think we talked about this in Slack already, but to make it explicit: Please open a new issue and describe your situation. This error message can have a million reasons really, and I doubt you have the same problem as Omar. Thanks.

**Comment by Sytten** [comment]
Many issues have been fixed in more recent prisma releases (>2.20 at least), see https://github.com/prisma/prisma/issues/6329#issuecomment-811097812.

**Comment by albertoperdomo** [comment]
Hello @omar-dulaimi and others,

we are working on a Prisma Data Proxy which will help with connection management in serverless backens. If you are interested in our upcoming Early Acess, you can sign up here: https://pris.ly/prisma-data-proxy

**Comment by omar-dulaimi** [comment]
Hello @albertoperdomo and thank you for letting me know. Unfortunately, we have moved away from serverless as our main compute environments.

**Comment by garrensmith** [comment]
This is fixed. The full information can be found here https://github.com/importcjj/mobc/issues/63
This will be out in the next release. You can test it now with prisma@dev
